### PR TITLE
Retrieve CAPI/CAPA from release image

### PR DESCRIPTION
**Pull request by bryan-cox**
**What this PR does / why we need it**:
Retrieves the CAPI and CAPA image components from the release image rather than registry.ci.openshift.org/hypershift/cluster-api:v1.0.0 and registry.ci.openshift.org/hypershift/cluster-api-aws-controller:v1.1.0 respectively

**Which issue(s) this PR fixes**:
Fixes [HOSTEDCP-577](https://issues.redhat.com/browse/HOSTEDCP-577)

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

File: hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go [hunk 1]
```diff
@@ -334,6 +351,21 @@ func reconcileAWSCluster(awsCluster *capiawsv1.AWSCluster, hcluster *hyperv1.Hos
 	return nil
 }
 
+// getPayloadImage get an image from the payload for a particular component.
+func getPayloadImage(ctx context.Context, imageRef string, pullSecret []byte, component string) (string, error) {
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go:355]: can we move the existing func into support/util/imagemetadata and reuse it?

File: hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go [hunk 2]
```diff
@@ -63,7 +63,24 @@ func (p AWS) ReconcileCAPIInfraCR(ctx context.Context, c client.Client, createOr
 }
 
 func (p AWS) CAPIProviderDeploymentSpec(hcluster *hyperv1.HostedCluster, hcp *hyperv1.HostedControlPlane) (*appsv1.DeploymentSpec, error) {
-	providerImage := imageCAPA
+	var pullSecret corev1.Secret
+	ctx := context.TODO()
+	c, err := cmdUtil.GetClient()
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go:68]: We have a pattern already to pass images
see
```
type AWS struct {
	utilitiesImage string
}
```

Let's add a field for the caprovider images or consider passing the image as a parameter in the interface signature

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 3]
```diff
@@ -159,6 +159,24 @@ func (r *NodePoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from nodepool: %w", err)
 			}
 		}
+
+		//Ensure all machines are gone before deleting related machine secrets
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:163]: this logic is already existing in `r.delete()` see `listSecrets` and `deleteMachineSet()`, please handle order in `r.delete()` if needed.

Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:163]: I could be missing it, but I don't think r.delete() is handling this. The issue seemed to be that the userSecret needs to stay around until all of the machine sets in the nodepool being deleted are truly gone. I thought the logic in r.delete() was just deleting a single machine set.

I can move this logic to r.delete() though.

File: hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go (hunk 2 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go]: Not sure what you mean here, but did my recent changes make this comment OBE? 

If not, would you please explain a little further so I can understand what's being asked?

**Comment by sjenning** [comment]
This change is causing a huge number of GETs from the hypershift-operator
```
=== RUN   TestEtcdChaos/EnsureAPIBudget/hypershift-operator_read
    util.go:604: over budget: budget: 5000, actual: 100247
```

**Comment by bryan-cox** [comment]
> This change is causing a huge number of GETs from the hypershift-operator
> 
> ```
> === RUN   TestEtcdChaos/EnsureAPIBudget/hypershift-operator_read
>     util.go:604: over budget: budget: 5000, actual: 100247
> ```

I think this might be due to the change I made in order to delete the AWS machine sets. I'll try reworking that and see if that resolves the problem.

File: hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go (hunk 2 above)
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go]: rather can enable a function that create a client everytime, let's reuse our shared existing instantiated once only clients. We what we do with utilitiesImage in New, you could do the same with the caprovider images and let getPlatform to find the image for each platform

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 3 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go]: I was missing how deleteMachineSet() worked but I understand better now. 

I have the deleteSecrets commented out for now because if the secrets are deleted prior to the machines being fully deleted, it causes this error in the capi controller, `failed to retrieve bootstrap data secret for AWSMachine clusters-brcox-hypershift-x86/brcox-hypershift-x86-work-1-glht8: Secret \"user-data-brcox-hypershift-x86-work-1-12f2ebfe\" not found" `. Moving `deleteNodePoolSecrets` to the end of `delete` does not solve the issue either.

File: hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go (hunk 2 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go]: I'm sorry. I understand more but I don't think all the way. Are you suggesting to look up the capiProviderImage within `hostedcluster_controller.go` and then pass it to `platform.GetPlatform(...)`?
<br/>

> let getPlatform to find the image for each platform

This is the part I'm confused on. I thought you were suggesting to look up the capiProviderImage within `platform.go` and reuse `GetPayloadImage` and `getPullSecretFromHostedCluster` from the updated `imagemetadata.go` in this PR but I also thought you were asking me to avoid using `cmdUtil.GetClient()`, which I do in `getPullSecretFromHostedCluster`.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 3 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go]: In additional to what is implemented in the PR currently, I also have this working locally at the `nodepool_controller.go` level:
```
if nodePool.Spec.Platform.AWS != nil {
	allMachinesTerminated := false
	allMachinesTerminated, err := r.allMachinesTerminated(nodePool)
	if err != nil {
		return ctrl.Result{}, err
	}

	if !allMachinesTerminated {
		return ctrl.Result{Requeue: true}, nil
	}

	r.deleteNodePoolSecrets(ctx, nodePool)
}

// Now we can remove the finalizer.
if controllerutil.ContainsFinalizer(nodePool, finalizer) {
	controllerutil.RemoveFinalizer(nodePool, finalizer)
	if err := r.Update(ctx, nodePool); err != nil {
		return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from nodepool: %w", err)
	}
}
log.Info("Deleted nodepool", "name", req.NamespacedName)
```

I'm open to suggestions or other paths, but basically it seems like we can't delete the NodePool secrets until all the NodePool machines are gone.

File: hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go (hunk 2 above)
Comment (enxebre) [reply hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go]: Pleased don't use cmdUtil.GetClient() within a func. 

>@sjenning I think this might be due to the change I made in order to delete the AWS machine sets. I'll try reworking that and see if that resolves the problem.

The recreated every loop non cached client is most likely causing that.

>Are you suggesting to look up the capiProviderImage within hostedcluster_controller.go and then pass it to platform.GetPlatform(...)

Either that or doing it from within `GetPlatform` func itself, there's many choices.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 4]
```diff
@@ -901,16 +904,34 @@ func (r *NodePoolReconciler) delete(ctx context.Context, nodePool *hyperv1.NodeP
 		return fmt.Errorf("failed to delete ConfigMaps with nodePool label: %w", err)
 	}
 
-	if err := deleteMachineDeployment(ctx, r.Client, md); err != nil {
-		return fmt.Errorf("failed to delete MachineDeployment: %w", err)
+	if nodePool.Spec.Platform.AWS != nil {
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:907]: Please let's drop this code. We dont' want to do an specific platform check here, nor rely on sleep.

What are you trying to achieve with this logic?

Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:907]: I was trying to prevent the NodePool secrets from being deleted before the machines are fully deleted. It’s related to this comment - https://github.com/openshift/hypershift/pull/1826#discussion_r1006055749

File: hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go (hunk 2 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/hostedcluster/internal/platform/aws/aws.go]: Ok will do

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 5]
```diff
@@ -2017,3 +2037,48 @@ func (r *NodePoolReconciler) doesCPODecompressAndDecodeConfig(ctx context.Contex
 	_, managesDecompressAndDecodeConfig := supportutil.ImageLabels(controlPlaneOperatorImageMetadata)[controlPlaneOperatorManagesDecompressAndDecodeConfig]
 	return managesDecompressAndDecodeConfig, nil
 }
+
+func (r *NodePoolReconciler) allMachinesTerminated(nodePool *hyperv1.NodePool) (bool, error) {
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:2041]: please drop this, see https://github.com/openshift/hypershift/pull/1826/files#r1007344121

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 4 above)
Comment (enxebre) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:907]: I see, so please paste a link with the diff with the change in capa that requires to do this now.
Let's fix it upstream. And here for now let's gate on Machines being gone.

Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:907]: This change seems to have come from [PR 2271](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2271). Specifically from the [changes](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2271/files#diff-4f8240f4e7ff07dcfa9627342b62772c1fe6ae0943021abe53fbb966db88886e) in machine.go in `UseSecretsManager ` and the [changes](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2271/files#diff-0b559bbd149f0e6d54d789235423f66fa8906fdc6ee9c99b9e85db912912011e) in awsmachine_controller.go in `reconcileDelete`.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 5 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:2041]: I'm assuming this function can stay for now as part of gating that all the Machines are gone per - https://github.com/openshift/hypershift/pull/1826#discussion_r1007374630

Comment (enxebre) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:2041]: We don't need anything of this nor the loop above.
All we need is a func deleteMachines that short circuit the workflow until there's not Machines persisted in etcd. e.g listMachines == 0

**Comment by enxebre** [comment]
@bryan-cox let's move forward with and ship this:
>This change seems to have come from https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2271. Specifically from the [changes](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2271/files#diff-4f8240f4e7ff07dcfa9627342b62772c1fe6ae0943021abe53fbb966db88886e) in machine.go in UseSecretsManager and the [changes](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2271/files#diff-0b559bbd149f0e6d54d789235423f66fa8906fdc6ee9c99b9e85db912912011e) in awsmachine_controller.go in reconcileDelete.

Can you share logs from capa you are having when trying to delete the secret before the awsMachine?

Let's address https://github.com/openshift/hypershift/pull/1826#discussion_r1009210462
then should be good to go.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 4 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:907]: Forgot to add an example of the log messages. The following errors are observed when the secret is deleted before the AWS machine has fully terminated:
```
I1019 13:05:07.838355 1 awsmachine_controller.go:287] "msg"="Handling deleted AWSMachine"
I1019 13:05:07.838456 1 awsmachine_controller.go:763] "msg"="Deleting unneeded entry from AWS S3" "secretPrefix"=""
I1019 13:05:07.838473 1 awsmachine_controller.go:287] "msg"="Handling deleted AWSMachine"
E1019 13:05:07.838499 1 awsmachine_controller.go:292] "msg"="unable to delete machine" "error"="failed to retrieve bootstrap data secret for AWSMachine clusters-brcox-hypershift-x86/brcox-hypershift-x86-us-east-1a-566vx: Secret \"user-data-brcox-hypershift-x86-us-east-1a-5c8b7c39\" not found"
I1019 13:05:07.838557 1 awsmachine_controller.go:763] "msg"="Deleting unneeded entry from AWS S3" "secretPrefix"=""
E1019 13:05:07.838589 1 awsmachine_controller.go:292] "msg"="unable to delete machine" "error"="failed to retrieve bootstrap data secret for AWSMachine clusters-brcox-hypershift-x86/brcox-hypershift-x86-us-east-1a-bsx4q: Secret \"user-data-brcox-hypershift-x86-us-east-1a-5c8b7c39\" not found"
E1019 13:05:07.839019 1 controller.go:317] controller/awsmachine "msg"="Reconciler error" "error"="failed to retrieve bootstrap data secret for AWSMachine clusters-brcox-hypershift-x86/brcox-hypershift-x86-us-east-1a-566vx: Secret \"user-data-brcox-hypershift-x86-us-east-1a-5c8b7c39\" not found" "name"="brcox-hypershift-x86-us-east-1a-566vx" "namespace"="clusters-brcox-hypershift-x86" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="AWSMachine"
I1019 13:05:07.839079 1 recorder.go:103] events "msg"="Warning" "message"="failed to retrieve bootstrap data secret for AWSMachine clusters-brcox-hypershift-x86/brcox-hypershift-x86-us-east-1a-566vx: Secret \"user-data-brcox-hypershift-x86-us-east-1a-5c8b7c39\" not found" "object"={"kind":"AWSMachine","namespace":"clusters-brcox-hypershift-x86","name":"brcox-hypershift-x86-us-east-1a-566vx","uid":"c0014da8-985d-481d-92bf-d6d7fb290e4a","apiVersion":"infrastructure.cluster.x-k8s.io/v1beta1","resourceVersion":"8072491"} "reason"="FailedGetBootstrapData"
I1019 13:05:07.839119 1 recorder.go:103] events "msg"="Warning" "message"="failed to retrieve bootstrap data secret for AWSMachine clusters-brcox-hypershift-x86/brcox-hypershift-x86-us-east-1a-566vx: Secret \"user-data-brcox-hypershift-x86-us-east-1a-5c8b7c39\" not found" "object"={"kind":"AWSMachine","namespace":"clusters-brcox-hypershift-x86","name":"brcox-hypershift-x86-us-east-1a-566vx","uid":"c0014da8-985d-481d-92bf-d6d7fb290e4a","apiVersion":"infrastructure.cluster.x-k8s.io/v1beta1","resourceVersion":"8072491"} "reason"="FailedGetBootstrapData"
...
```

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 5 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go]: I believe this has been addressed now.

**Comment by bryan-cox** [comment]
> @bryan-cox let's move forward with and ship this:
> 
> > This change seems to have come from [kubernetes-sigs/cluster-api-provider-aws#2271](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2271). Specifically from the [changes](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2271/files#diff-4f8240f4e7ff07dcfa9627342b62772c1fe6ae0943021abe53fbb966db88886e) in machine.go in UseSecretsManager and the [changes](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2271/files#diff-0b559bbd149f0e6d54d789235423f66fa8906fdc6ee9c99b9e85db912912011e) in awsmachine_controller.go in reconcileDelete.
> 
> Can you share logs from capa you are having when trying to delete the secret before the awsMachine?
> 
> Let's address [#1826 (comment)](https://github.com/openshift/hypershift/pull/1826#discussion_r1009210462) then should be good to go.
> 
> Orthogonally [kubernetes-sigs/cluster-api-provider-aws#3805](https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/3805)

Ok I think I've addressed  [#1826 (comment)](https://github.com/openshift/hypershift/pull/1826#discussion_r1009210462)

I included a snippet of the error log messages in [here](https://github.com/openshift/hypershift/pull/1826#discussion_r1015430528)

File: support/util/imagemetadata.go [hunk 6]
```diff
@@ -92,3 +94,18 @@ func ImageLabels(metadata *dockerv1client.DockerImageConfig) map[string]string {
 		return metadata.ContainerConfig.Labels
 	}
 }
+
+// GetPayloadImage get an image from the payload for a particular component
+func GetPayloadImage(ctx context.Context, hc *hyperv1.HostedCluster, component string, pullSecret []byte) (string, error) {
+	riprovider := &releaseinfo.RegistryClientProvider{}
```
Comment (enxebre) [inline support/util/imagemetadata.go:100]: riprovider?

File: hypershift-operator/controllers/hostedcluster/hostedcluster_webhook_test.go [hunk 7]
```diff
@@ -64,7 +69,7 @@ func TestWebhookAllowsHostedClusterReconcilerUpdates(t *testing.T) {
 				},
 				&corev1.Secret{
 					ObjectMeta: metav1.ObjectMeta{Namespace: "some-ns"},
-					Data:       map[string][]byte{".dockerconfigjson": []byte("something")},
+					Data:       map[string][]byte{".dockerconfigjson": []byte("{\"something\": \"something\"}")},
```
Comment (enxebre) [inline hypershift-operator/controllers/hostedcluster/hostedcluster_webhook_test.go:72]: why is this change needed?

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 8]
```diff
@@ -2017,3 +2032,48 @@ func (r *NodePoolReconciler) doesCPODecompressAndDecodeConfig(ctx context.Contex
...
+
+func (r *NodePoolReconciler) ensureMachineDeletion(nodePool *hyperv1.NodePool) error {
+	machineSetList := &unstructured.UnstructuredList{}
+	var gvk schema.GroupVersionKind
+	var err error
+
+	switch nodePool.Spec.Platform.Type {
+	// Define the desired template type and mutateTemplate function.
+	case hyperv1.AWSPlatform:
+		gvk, err = apiutil.GVKForObject(&capiaws.AWSMachineList{}, api.Scheme)
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:2044]: we don't need anything of this. All we need to do is check there's no capi.Machines in the hcp namespace.

File: support/util/imagemetadata.go (hunk 6 above)
Comment (bryan-cox) [reply support/util/imagemetadata.go:100]: I spelled it out more - releaseImageProvider.

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 8 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:2044]: Oh ok, I thought I would need to look at the specific CAPA instance to get the machines. Making the change now and testing.

File: support/releaseinfo/registryclient/client.go [hunk 9]
```diff
@@ -234,7 +234,7 @@ func ExtractImageFilesToDir(ctx context.Context, imageRef string, pullSecret []b
 func getMetadata(ctx context.Context, imageRef string, pullSecret []byte) ([]distribution.Descriptor, distribution.BlobStore, error) {
 	repo, ref, err := GetRepoSetup(ctx, imageRef, pullSecret)
 	if err != nil {
-		return nil, nil, fmt.Errorf("failed to repo setup: %w", err)
+		return nil, nil, fmt.Errorf("failed to setup repo: %w", err)
```
Comment (enxebre) [inline support/releaseinfo/registryclient/client.go:237]: nit: failed to get repo setup

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 10]
```diff
@@ -2017,3 +2032,22 @@ func (r *NodePoolReconciler) doesCPODecompressAndDecodeConfig(ctx context.Contex
...
+func (r *NodePoolReconciler) ensureMachineDeletion(nodePool *hyperv1.NodePool, controlPlaneNamespace string) error {
+	machines := capiv1.MachineList{}
+	if err := r.List(context.Background(), &machines, &client.ListOptions{Namespace: controlPlaneNamespace}); err != nil {
+		return fmt.Errorf("failed to list MachineSet: %w", err)
+	}
+
+	// Filter out only machines belonging to deleted NodePool
+	var machineSetOwnedMachines []capiv1.Machine
+	for i, machine := range machines.Items {
+		if machine.Annotations[nodePoolAnnotation] == "clusters/"+nodePool.GetName() {
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:2045]: this name is not guaranteed "clusters" is your particular ns

Comment (enxebre) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:2045]: You could also check for existence of
machineDeployment(nodePool *hyperv1.NodePool, controlPlaneNamespace string)
or
machineSet(nodePool *hyperv1.NodePool, controlPlaneNamespace string)
And then filter machines by ownerRef.

File: support/releaseinfo/registryclient/client.go (hunk 9 above)
Comment (bryan-cox) [reply support/releaseinfo/registryclient/client.go]: Updated message

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 10 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:2045]: Understood about clusters not being a guarantee. I will looking into changing this tomorrow.

I sent a Slack message in a thread, but when I was testing, I couldn't look at the ownerRef because the MachineSet name returned [here](https://github.com/bryan-cox/hypershift/blob/910c2419e355f2564eff0fc52cb87c78b3862fff/hypershift-operator/controllers/nodepool/nodepool_controller.go#L873) is not the actual MachineSet name so then I can't compare the ownerRef. It didn't seem like the MachineSet name is exactly identical to the NodePool name. 

For example, when I was testing, my NodePool name was _brcox-hypershift-x86-test-us-east-1a_ but the MachineSet name was _brcox-hypershift-x86-test-us-east-1a-bff454f9f_. I was going to also test tomorrow to see if this [function](https://github.com/bryan-cox/hypershift/blob/910c2419e355f2564eff0fc52cb87c78b3862fff/hypershift-operator/controllers/nodepool/nodepool_controller.go#L810) is even working based on this.

Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go]: I think I found a way that works now.

Comment (enxebre) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go]: >I couldn't look at the ownerRef because the MachineSet name returned [here](https://github.com/bryan-cox/hypershift/blob/910c2419e355f2564eff0fc52cb87c78b3862fff/hypershift-operator/controllers/nodepool/nodepool_controller.go#L873)

That's because that MachineSet is the one created explicitly for one you set in-place upgrade typer.

You need to account for both scenarios: replace (MachineDeployment known name - which generates machineSets underneath)
and inplace (MachineSet known name)

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 11]
```diff
@@ -901,16 +903,30 @@ func (r *NodePoolReconciler) delete(ctx context.Context, nodePool *hyperv1.NodeP
 		return fmt.Errorf("failed to delete ConfigMaps with nodePool label: %w", err)
 	}
 
-	if err := deleteMachineDeployment(ctx, r.Client, md); err != nil {
-		return fmt.Errorf("failed to delete MachineDeployment: %w", err)
+	// Ensure all machines in NodePool are deleted
+	err = r.ensureMachineDeletion(nodePool, controlPlaneNamespace)
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:907]: nit: this can be inlined

File: hypershift-operator/controllers/nodepool/nodepool_controller.go [hunk 12]
```diff
@@ -2017,3 +2032,27 @@ func (r *NodePoolReconciler) doesCPODecompressAndDecodeConfig(ctx context.Contex
...
+
+	// Filter out only machines belonging to deleted NodePool
+	var machineSetOwnedMachines []capiv1.Machine
+	for i, machine := range machines.Items {
+		if machine.Annotations[nodePoolAnnotation] == client.ObjectKeyFromObject(nodePool).String() {
+			machineSetOwnedMachines = append(machineSetOwnedMachines, machines.Items[i])
+		}
+	}
+	if len(machineSetOwnedMachines) > 0 {
+		return fmt.Errorf("")
```
Comment (enxebre) [inline hypershift-operator/controllers/nodepool/nodepool_controller.go:2055]: add message to return e.g. fmt.Errorf("there are still Machines in the cluster")

**Comment by enxebre** [comment]
Dropped 2 more nits.
This is great, thanks @bryan-cox!
/approve
will ship after https://github.com/openshift/hypershift/pull/1868

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 12 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:2055]: Ok, it's been updated

File: hypershift-operator/controllers/nodepool/nodepool_controller.go (hunk 11 above)
Comment (bryan-cox) [reply hypershift-operator/controllers/nodepool/nodepool_controller.go:907]: Ok, it's been updated

**Comment by enxebre** [comment]
This seems to be failing consistently @bryan-cox, did you have the chance to test it?

**Comment by enxebre** [comment]
This PR keeps failing leaking resources, have you seen it working locally?
Can you please bump the timeout for the PR here https://github.com/openshift/hypershift/blob/9878207a11d99c2280f438d85e85311b7b90ae94/test/e2e/util/fixture.go#L290
If that still fails then failure is legit.

**Comment by eranco74** [comment]
The job that tests ocm-2.6 (ocp 4.11) is still trying to use registry.ci.openshift.org/hypershift/cluster-api:v1.0.0.
For some reason, the above image no longer exists (since Jan 25)
/cherry-pick release-4.11
